### ISSUE TITLE
CompositePresentationEvent

### ISSUE THREAD

**Issue by isvaljek** [PrismLibrary/Prism#152]
What is the replacement for CompositePresentationEvent? How to share EventAggregator between different ViewModel instances?
[This one](http://blog.magnusmontin.net/2014/02/28/using-the-event-aggregator-pattern-to-communicate-between-view-models/) uses EventAggregator through constructor injection, but then I could have more easily injected my shared object that way.

**Comment by briannoyes** [comment]
EventAggregator is still in Prism 6, it has just moved into the Prism.Core PCL library. The event type is PubSubEvent instead of CompositePresentationEvent.

Just like before, a singleton instance of EventAggregator is registered in the container by the bootstrapper, so constructor injection is the recommended way of getting a reference to it.

**Comment by isvaljek** [comment]
Does that mean I only need to define a VM constructor that accepts the EventAggregator and it will automagically be passed along (no need for literal pass)?
Also, I have one ViewModel that already has two parameters, how can I squeeze in this third?
Thanks.

**Comment by briannoyes** [comment]
As long as the VM is being constructed through the container, which it will be if you either use Region navigation or you use ViewModelLocator. You can have as many constructor parameters as your ViewModel needs, as long as the container has registrations for the types of those parameters. 

If you were manually constructing your VM and passing in those parameters yourself, then you would have to go with a "poor man's" DI approach - expose a static reference to the container somewhere - i.e. Bootstrapper which can be populated in the CreateContainer override, and then use the static ref to programmatically Resolve event aggregator.

**Comment by isvaljek** [comment]
Is the PubSub the right thing for the app if I have a three level hierarchy of views/viewmodels and only want to get an object property of the top level ViewModel to be shared/accessible between ViewModels?

**Comment by briannoyes** [comment]
You could use that. What I would recommend instead is a client service or repository that manages that shared state and that both the low level and higher level VM's inject as a dependency. That repository can raise normal .NET events when the shared data changes so that the other VM gets passed the data that way.

**Comment by isvaljek** [comment]
But what if the 'lower' View/Viewmodels are created dynamically? When the top level viewmodel changes the state and that raises an event, the lower level viewmodels won't receive it (since they don't exist yet), right?
I'm a bit lost here, maybe too much php/javascript lately, can you suggest some good reading on Prism 6 (that has sharing/dependency injection covered)?

**Comment by briannoyes** [comment]
I would try to handle that by setting up a pattern in the lower view models that they inject the repository when they are constructed, immediately subscribe for state changes, and retrieve current values as part of construction. Then as long as the instance sticks around, they will get the event that tells them the state has changed, and each time they are constructed new they retrieve the current state.

In terms of reading or tutorials, the book we wrote on Prism 4 is still accurate for Prism 6 with respect to most things about modularity, regions, commands, and events other than the namespaces have changed. We added ViewModelLocator and BindableBase and a few other MVVM related things in Prism 5 and then enhanced some more in Prism 6. The latest official docs before they handed over the code to the community is here:
https://msdn.microsoft.com/en-us/library/gg406140.aspx

Brian Lagunas and I have some courses on Prism in the Pluralsight library as well. If you do not have a subscription shoot me a note brian.noyes at gmail.com and I can get you a 1 month pass.
